fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/ConcreteAvailability.java

```diff
@@ -0,0 +1,90 @@
+// Copyright 2017 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.table.availability;
+
+import com.yahoo.bard.webservice.data.config.names.TableName;
+import com.yahoo.bard.webservice.metadata.DataSourceMetadataService;
+import com.yahoo.bard.webservice.table.Column;
+import com.yahoo.bard.webservice.table.resolver.DataSourceConstraint;
+import com.yahoo.bard.webservice.util.IntervalUtils;
+import com.yahoo.bard.webservice.util.SimplifiedIntervalList;
+
+import org.joda.time.Interval;
+
+import avro.shaded.com.google.common.collect.ImmutableSet;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * An availability that provides column and table available interval services for concrete physical table.
+ */
+public class ConcreteAvailability implements Availability {
+
+    private final TableName name;
+    private final Set<Column> columns;
+    private final DataSourceMetadataService metadataService;
+
+    private final Set<String> columnNames;
+
+    /**
+     * Constructor.
+     *
+     * @param tableName  The name of the table and data source associated with this Availability
+     * @param columns  The columns associated with the table and availability
+     * @param metadataService  A service containing the datasource segment data
+     */
+    public ConcreteAvailability(
+            TableName tableName,
+            @NotNull Set<Column> columns,
+            @NotNull DataSourceMetadataService metadataService
+    ) {
+        this.name = tableName;
+        this.columns = ImmutableSet.copyOf(columns);
+        this.metadataService = metadataService;
+
+        this.columnNames = columns.stream().map(Column::getName).collect(Collectors.toSet());
+    }
+
+    @Override
+    public Set<TableName> getDataSourceNames() {
+        return Collections.singleton(name);
```

Comment:
In almost every other place in the codebase, we've generally pushed conversions like this (ie. from an instance into an immutable collection due to a collection getter) into the constructor, rather than call it multiple times. So, from that standpoint, doing so maintains consistency with the codebase.

The other perspective on this is that a `ConcreteAvailability` will generally live for the life of the system, and so this method will be called at least once per request. While I'm sure it's not a huge impact, creating more objects than needed causes more garbage overhead, which introduces less consistency in the performance of the system. So, having more performant code (especially in terms of consistency), for _slightly_ less maintainability, is likely a worthwhile reason.  In theory, this is something the JIT _should_ be able to optimize away, so if you can show that the JIT _will_ optimize the repeated object away, I'm all for letting the JIT do the work.

Lastly, while your point about having a collection as a private field _could_ be slightly confusing, a good name (perhaps like `cachedDataSourceNames`) mitigates that concern significantly.